storage: better debug output for consistency checker

vivekmenezes opened the pull request:
This is to nail down any consistency check failure we might find.

closes #5700 

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/5403)

<!-- Reviewable:end -->

petermattis commented on storage/replica_command.go:
```diff
@@ -1477,29 +1480,38 @@ func (r *Replica) sha512(desc roachpb.RangeDescriptor, snap engine.Engine) ([]by
 		key := iter.Key()
 		value := iter.Value()
+
+		if log.V(1) {
```
Rather than test `log.V(1)` inside the loop, I think it would be better to initialize `debug` based on `log.V(1)` outside the loop and change the above test to `if debug != nil {`. The `debug` initialization would look like:

```
  var debug *roachpb.VerifyChecksumRequest_Debug
  if log.V(1) {
    debug = &roachpb.VerifyChecksumRequest_Debug{}
  }
```

Then at the end of this method you can just return `debug` without having to zero it if `!log.V(1)`.

tamird commented:
Reviewed 4 of 6 files at r1.
Review status: 4 of 6 files reviewed at latest revision, 2 unresolved discussions.

---

*[roachpb/api.proto, line 559 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KDGR7JhTjQUogZffJL4:-KDGR7JhTjQUogZffJL5:-150787670) ([raw file](https://github.com/cockroachdb/cockroach/blob/47f82325faaa6329092d8c2a7d86ae3abf557049/roachpb/api.proto#L559)):*
this message already exists in data.proto. you can replace this entire addition with
```
repeated debug_entires KeyValue = 1 [(gogoproto.nullable) = false];
```

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403)*
<!-- Sent from Reviewable.io -->

petermattis commented on storage/replica_command.go:
```diff
@@ -1532,6 +1544,15 @@ func (r *Replica) VerifyChecksum(batch engine.Engine, ms *engine.MVCCStats, h ro
+			for i := range debug.Entries {
+				e := debug.Entries[i]
+				v := c.debug.Entries[i]
+				if !bytes.Equal(e.Key, v.Key) || !bytes.Equal(e.Value, v.Value) {
```
The output is going to be very verbose when the keys don't match. Might be better to use something like `github.com/kr/pretty.Diff` here which should output a minimal difference, though I'm not sure if we're linking that library into the production binary yet.

bdarnell commented:
Review status: 4 of 6 files reviewed at latest revision, 6 unresolved discussions.

---

*[roachpb/api.proto, line 559 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KDGR7JhTjQUogZffJL4:-KDJPPqy-8p9zJ9cw-mO:-871980585) ([raw file](https://github.com/cockroachdb/cockroach/blob/47f82325faaa6329092d8c2a7d86ae3abf557049/roachpb/api.proto#L559)):*
Or `RaftSnapshotData`, which is designed to include everything that matters for consistency across replicas.

---

*[roachpb/api.proto, line 561 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KDJPKJ6iDHR5_t6mEhf:-KDJPKJ7vwBppiys9JqI:-1030820678) ([raw file](https://github.com/cockroachdb/cockroach/blob/47f82325faaa6329092d8c2a7d86ae3abf557049/roachpb/api.proto#L561)):*
Timestamps can cause checksum mismatches too. 

---

*[storage/replica_command.go, line 1549 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KDJPk6EXQMtoP4YDWyV:-KDJPk6EXQMtoP4YDWyW:476307580) ([raw file](https://github.com/cockroachdb/cockroach/blob/47f82325faaa6329092d8c2a7d86ae3abf557049/storage/replica_command.go#L1549)):*
`debug` may be nil, in which case this will panic. This is also why it may be a bad idea to control non-logging behavior with a verbosity flag: it gives us an opportunity to write code that works in CI (where we turn on V(1) globally) but fails in production. I think this feature should be off by default (even in CI).

---

*[storage/replica_command.go, line 1551 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KDJQnRYNgYKi1Mvd45i:-KDJQnRZSSXKh3uUU48b:-1002517041) ([raw file](https://github.com/cockroachdb/cockroach/blob/47f82325faaa6329092d8c2a7d86ae3abf557049/storage/replica_command.go#L1551)):*
This will panic if the two debug lists have different numbers of entries. If an entry is inserted or removed then every subsequent entry will report a mismatch.

---

*[storage/replica_command.go, line 1552 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KDFKFZ7QUAQwa1KaHqk-r1-1552:-KDJRIbFchZCuUshTDcZ:799936285) ([raw file](https://github.com/cockroachdb/cockroach/blob/47f82325faaa6329092d8c2a7d86ae3abf557049/storage/replica_command.go#L1552)):*
Or we could dump everything in a form that the user can run `diff` or `cmp` on. 

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 6 files at r1, 1 of 1 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 0 of 13 files reviewed at latest revision, 7 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KE9neEblMFpKHvqmhO0:-KE9neEblMFpKHvqmhO1:697555152) (no related file):*
This is a born again PR. Please take a fresh look at it.

---

*[roachpb/api.proto, line 559 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KDGR7JhTjQUogZffJL4:-KE9nGVQtbl-AG0te1lK:1658755450) ([raw file](https://github.com/cockroachdb/cockroach/blob/47f82325faaa6329092d8c2a7d86ae3abf557049/roachpb/api.proto#L559)):*
I've used KeyValue because it's a lot easier with the dependencies. Including a Raft message in api.proto is not straightforward.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 7 files at r4, 12 of 12 files at r5.
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[acceptance/cluster/cluster.go, line 66 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KEBFcEFTNpgU8ZJhisz:-KEBFcEFTNpgU8ZJhit-:-1984294713) ([raw file](https://github.com/cockroachdb/cockroach/blob/66db1b7c310585d1f7b583aed08032adcac3ea64/acceptance/cluster/cluster.go#L66)):*
this was probably supposed to be in the first commit

---

*[kv/dist_sender_server_test.go, line 454 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KEBFkOnNuODlg6zAdm3:-KEBFkOnNuODlg6zAdm4:166874) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ec639ff6140318514372f5485481097f4e8f15/kv/dist_sender_server_test.go#L454)):*
?

---

*[roachpb/api.proto, line 559 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KDGR7JhTjQUogZffJL4:-KEBFoBQlDdNosV9ZqMf:-1479526467) ([raw file](https://github.com/cockroachdb/cockroach/blob/47f82325faaa6329092d8c2a7d86ae3abf557049/roachpb/api.proto#L559)):*
RaftSnapshotData is in `etcd/raft/raftpb/raft.proto`, you should have no problem including it here.

---

*[storage/replica_command.go, line 1657 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KEBNVkIm7ZtsSNeteiN:-KEBNVkJeHanSjdYqFbo:-460357658) ([raw file](https://github.com/cockroachdb/cockroach/blob/66db1b7c310585d1f7b583aed08032adcac3ea64/storage/replica_command.go#L1657)):*
why is this sound? this implementation also mishandles missing keys in either leader or replica

---

*[storage/store.go, line 139 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KEBGr82EAuTEvrsEFiG:-KEBGr82EAuTEvrsEFiH:166874) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ec639ff6140318514372f5485481097f4e8f15/storage/store.go#L139)):*
?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403)*
<!-- Sent from Reviewable.io -->

vivekmenezes committed:
roachpb: remove internal_raft.proto -> api.proto dependency

bdarnell commented:
Reviewed 1 of 7 files at r4, 12 of 12 files at r5.
Review status: all files reviewed at latest revision, 16 unresolved discussions, some commit checks failed.

---

*[roachpb/api.proto, line 559 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KDGR7JhTjQUogZffJL4:-KEBtnYHzv9YHFjTLoaA:-1869241972) ([raw file](https://github.com/cockroachdb/cockroach/blob/47f82325faaa6329092d8c2a7d86ae3abf557049/roachpb/api.proto#L559)):*
The problem is that `api.proto` is used from c++ and importing it here would require shaving some yaks in the c++ proto build. Using `KeyValue` here LGTM.

---

*[roachpb/api.proto, line 568 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KEBuNuxmxUPlV_C2Je4:-KEBuNuxmxUPlV_C2Je5:188123344) ([raw file](https://github.com/cockroachdb/cockroach/blob/66db1b7c310585d1f7b583aed08032adcac3ea64/roachpb/api.proto#L568)):*
Why 1001? We don't generally leave gaps in our tag numbers. (and if you're wanting to reserve some small tag numbers for space optimizations, I'd say that this field deserves a small tag number as much as anything because when it is used it's repeated many times).

---

*[storage/replica_command.go, line 1373 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KEBzfzUaj-N_ZTPot_q:-KEBzfzVbCiFv1ykv7aJ:834765100) ([raw file](https://github.com/cockroachdb/cockroach/blob/66db1b7c310585d1f7b583aed08032adcac3ea64/storage/replica_command.go#L1373)):*
In general, RPC methods should have all their arguments contained in the `args` proto; it's weird that `withDiff` is a separate argument that can only be controlled when bypassing the rpc path. 

While it's not new to this PR, it's also weird that the descriptor is passed in instead of using `r.Desc()`.

---

*[storage/replica_command.go, line 1614 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KEBygGGUbp-_bq1rj7o:-KEBygGH1qGbcvkLRLWa:1520020896) ([raw file](https://github.com/cockroachdb/cockroach/blob/66db1b7c310585d1f7b583aed08032adcac3ea64/storage/replica_command.go#L1614)):*
If we're not in panic mode, this could be expensive as each consistency check triggers one (or more, for each inconsistent replica) followup checks and dumps lots of data into the logs. In non-panic mode I would keep the logging the same as before and only do this second check in panic mode.

---

*[storage/replica_command.go, line 1639 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KEC-A79csmx4VCBYHPU:-KEC-A79csmx4VCBYHPV:1075018874) ([raw file](https://github.com/cockroachdb/cockroach/blob/66db1b7c310585d1f7b583aed08032adcac3ea64/storage/replica_command.go#L1639)):*
This function needs a test (it could return a list of strings instead of logging them directly).

---

*[storage/replica_command.go, line 1675 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KEC-fv20aPHCf2H3Zbe:-KEC-fv20aPHCf2H3Zbf:235288790) ([raw file](https://github.com/cockroachdb/cockroach/blob/66db1b7c310585d1f7b583aed08032adcac3ea64/storage/replica_command.go#L1675)):*
This doesn't look right: there may be many values with different timestamps for the same key. The `KeyValue` proto puts the timestamp as a part of the value, but logically it's more like a part of the key. 

---

*[storage/store.go, line 139 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KEBGr82EAuTEvrsEFiG:-KEBwBjK0CotrvGEyzz-:1453655870) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ec639ff6140318514372f5485481097f4e8f15/storage/store.go#L139)):*
Yesterday Vivek and I were talking and I said that we should figure out why we had to use `LocalMax` instead of `KeyMin`, and I thought we should probably fix it so that it would work to pass `KeyMin`. But now I remember why this check needs to stay: our keyspace is not truly monolithic. There's a gap where the local keys are. So for now we should continue to use `LocalMax` instead of `KeyMin` for the starting point of the consistency checker. 

I think there's probably something cleaner we can do here (e.g. set `KeyMin` equal to `LocalMax` since it's rarely useful to actually talk about the empty key), but let's not tackle that now.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 16 unresolved discussions, some commit checks failed.

---

*[roachpb/api.proto, line 559 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KDGR7JhTjQUogZffJL4:-KEC0dhVyvVf4SPbzrIv:-974209405) ([raw file](https://github.com/cockroachdb/cockroach/blob/47f82325faaa6329092d8c2a7d86ae3abf557049/roachpb/api.proto#L559)):*
Heh, the yak shave is not so bad. I can help with this.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403)*
<!-- Sent from Reviewable.io -->

vivekmenezes committed:
storage: better debug output for consistency checker

When VerifyChecksum() sees a consistency checksum mismatch, instead
of panicking, it issues a second consistency check by calling
ConsistencyCheck().

ConsistencyCheck() is modified to send an entire copy of the
range by default to each replica in VerifyChecksum(), so that
VerifyChecksum() can log a proper diff with checksum mismatch.
VerifyChecksum can only panic if it receives a copy of the range
and can output out proper debug output before panicking.

vivekmenezes commented:
Review status: 0 of 13 files reviewed at latest revision, 16 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KE9neEblMFpKHvqmhO0:-KEIaaFnr_2p2ZONTrO9:-1999248811) (no related file):*
Althought I pushed this, I'd like to take another look at this before you review.

---

*[acceptance/cluster/cluster.go, line 66 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KEBFcEFTNpgU8ZJhisz:-KEEHEWWDuqMK3kWOU0J:520555434) ([raw file](https://github.com/cockroachdb/cockroach/blob/66db1b7c310585d1f7b583aed08032adcac3ea64/acceptance/cluster/cluster.go#L66)):*
removed for the change altogether

---

*[roachpb/api.proto, line 559 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KDGR7JhTjQUogZffJL4:-KEEHJPwJ6f3HZO1AsAO:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/47f82325faaa6329092d8c2a7d86ae3abf557049/roachpb/api.proto#L559)):*
done

---

*[roachpb/api.proto, line 561 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KDJPKJ6iDHR5_t6mEhf:-KEEHL6i6a-fniaWW8lE:-1221266418) ([raw file](https://github.com/cockroachdb/cockroach/blob/47f82325faaa6329092d8c2a7d86ae3abf557049/roachpb/api.proto#L561)):*
thanks

---

*[roachpb/api.proto, line 568 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KEBuNuxmxUPlV_C2Je4:-KEEHNj-LHSMiB_IZkbY:-467915377) ([raw file](https://github.com/cockroachdb/cockroach/blob/66db1b7c310585d1f7b583aed08032adcac3ea64/roachpb/api.proto#L568)):*
fixed

---

*[storage/replica_command.go, line 1484 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KDFKFZ7QUAQwa1KaHqk-r1-1484:-KEEHPxavyQlCqvXDXwa:-1751637291) ([raw file](https://github.com/cockroachdb/cockroach/blob/47f82325faaa6329092d8c2a7d86ae3abf557049/storage/replica_command.go#L1484)):*
Stuff is a lot more automatic now, with the consistency checker always logging a diff before it panics.

---

*[storage/replica_command.go, line 1552 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KDFKFZ7QUAQwa1KaHqk-r1-1552:-KEEHd3a8UEAp1-E5FNi:-1783891096) ([raw file](https://github.com/cockroachdb/cockroach/blob/47f82325faaa6329092d8c2a7d86ae3abf557049/storage/replica_command.go#L1552)):*
I still have an ad hoc implementation, but it's a lot better now.

---

*[storage/replica_command.go, line 1373 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KEBzfzUaj-N_ZTPot_q:-KEEdFC8b_ky0ymeI6MC:-467915377) ([raw file](https://github.com/cockroachdb/cockroach/blob/66db1b7c310585d1f7b583aed08032adcac3ea64/storage/replica_command.go#L1373)):*
fixed

---

*[storage/replica_command.go, line 1614 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KEBygGGUbp-_bq1rj7o:-KEEGDdIaZo8DnWgbtV4:1736000640) ([raw file](https://github.com/cockroachdb/cockroach/blob/66db1b7c310585d1f7b583aed08032adcac3ea64/storage/replica_command.go#L1614)):*
well, I think the plan is to always panic. We have an existing non-panic mode because we're still in Beta. I think it would be useful for a Beta customer to give us the entire log containing all the debug information. Right now, all we know is stuff fails on a specific replica.

---

*[storage/replica_command.go, line 1639 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KEC-A79csmx4VCBYHPU:-KEIaY5X7OfJtQNDwS0F:676635175) ([raw file](https://github.com/cockroachdb/cockroach/blob/66db1b7c310585d1f7b583aed08032adcac3ea64/storage/replica_command.go#L1639)):*
Done

---

*[storage/replica_command.go, line 1657 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KEBNVkIm7ZtsSNeteiN:-KEEd6ZQZMVgkZLd_Stg:-1248765936) ([raw file](https://github.com/cockroachdb/cockroach/blob/66db1b7c310585d1f7b583aed08032adcac3ea64/storage/replica_command.go#L1657)):*
Is this a problem?

---

*[storage/replica_command.go, line 1675 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KEC-fv20aPHCf2H3Zbe:-KEEdBcwFglzlTqsis52:-467915377) ([raw file](https://github.com/cockroachdb/cockroach/blob/66db1b7c310585d1f7b583aed08032adcac3ea64/storage/replica_command.go#L1675)):*
fixed

---

*[storage/store.go, line 139 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KEBGr82EAuTEvrsEFiG:-KEEH9cHgs4H6jrwYgfn:164655668) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ec639ff6140318514372f5485481097f4e8f15/storage/store.go#L139)):*
Removed this change completely. Thanks

---

*[kv/dist_sender_server_test.go, line 454 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KEBFkOnNuODlg6zAdm3:-KEEHHOi4kGtGdiybUAp:1374286819) ([raw file](https://github.com/cockroachdb/cockroach/blob/78ec639ff6140318514372f5485481097f4e8f15/kv/dist_sender_server_test.go#L454)):*
same. removed

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 0 of 13 files reviewed at latest revision, 16 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KE9neEblMFpKHvqmhO0:-KEJuV7VHVawNkhJPic8:-909749207) (no related file):*
Ready for another review. Thanks

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Reviewed 2 of 15 files at r6, 8 of 11 files at r7, 3 of 3 files at r8.
Review status: all files reviewed at latest revision, 12 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KE9neEblMFpKHvqmhO0:-KEOFLVQiixppNzbov8c:676834233) (no related file):*
LGTM

---

*[storage/replica_command.go, line 1654 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KEOCrsSNwxnKfcaQ1Ja:-KEOCrsTZQrKJ9Ne5s3t:-2139716765) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b3b8ee3212a66cf9609582421f44c4dc9833c08/storage/replica_command.go#L1654)):*
You shouldn't have to explicitly call PrettyPrint; it will happen automatically when you `%s` a `roachpb.Key`. And why isn't `d.Key` already a `roachpb.Key`?

---

*[storage/replica_test.go, line 4707 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KEOEgxVh3row5MQJoff:-KEOEgxW9Wsh8XvoDYMw:-1189820754) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b3b8ee3212a66cf9609582421f44c4dc9833c08/storage/replica_test.go#L4707)):*
There should be some differing values in here too. I see that you're adjusting timestamps below, but I think it would be better to just make the inputs a list of key/timestamp/value triples.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 15 files at r6, 7 of 11 files at r7, 1 of 3 files at r8.
Review status: all files reviewed at latest revision, 15 unresolved discussions.

---

*[client/batch.go, line 376 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KEOFoWZ5qjpP9svfbJX:-KEOFoWaBCgJFGeHG68V:348387678) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b3b8ee3212a66cf9609582421f44c4dc9833c08/client/batch.go#L376)):*
this is a different tense than the rest of this doc comment.

---

*[client/db.go, line 368 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KEOFtbdwn7hRS-iLt0A:-KEOFtbdwn7hRS-iLt0B:711262620) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b3b8ee3212a66cf9609582421f44c4dc9833c08/client/db.go#L368)):*
this is awkward when following the existing comment. perhaps "it logs a diff of all inconsistent keys when..."

---

*[roachpb/api.proto, line 559 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KDGR7JhTjQUogZffJL4:-KEOGYzHwNjdWk8MLO8a:-822982725) ([raw file](https://github.com/cockroachdb/cockroach/blob/47f82325faaa6329092d8c2a7d86ae3abf557049/roachpb/api.proto#L559)):*
looks like you still moved RaftCommand, which is not necessary.

---

*[roachpb/internal_raft.proto, line 30 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KEOGUcLJZlynoqJGqQZ:-KEOGUcLJZlynoqJGqQ_:831603178) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a85cc05b183e61e9224fcec7b8acfe4d09cfa1c/roachpb/internal_raft.proto#L30)):*
why did this move?

---

*[storage/client_raft_test.go, line 1564 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KEOGiI7voMgSfIBfNdP:-KEOGiI7voMgSfIBfNdQ:-1844718083) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b3b8ee3212a66cf9609582421f44c4dc9833c08/storage/client_raft_test.go#L1564)):*
why does this need to `sleep`? can it be expressed as `SucceedsSoon`?

---

*[storage/replica_command.go, line 1657 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KEBNVkIm7ZtsSNeteiN:-KEOHFmUSst6gFQu_CEl:-1995644922) ([raw file](https://github.com/cockroachdb/cockroach/blob/66db1b7c310585d1f7b583aed08032adcac3ea64/storage/replica_command.go#L1657)):*
isn't it? if there is a missing key the diff won't tell you.

---

*[storage/replica_command.go, line 1557 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KEOHr5nrAWnOs6Auos1:-KEOHr5nrAWnOs6Auos2:682611859) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b3b8ee3212a66cf9609582421f44c4dc9833c08/storage/replica_command.go#L1557)):*
this function probably needs a new comment and a rename.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 2 of 13 files reviewed at latest revision, 15 unresolved discussions.

---

*[client/batch.go, line 376 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KEOFoWZ5qjpP9svfbJX:-KETmuoTpOxV1vh9-A09:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b3b8ee3212a66cf9609582421f44c4dc9833c08/client/batch.go#L376)):*
done

---

*[client/db.go, line 368 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KEOFtbdwn7hRS-iLt0A:-KETnIgKKOZIlh6OWC3H:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b3b8ee3212a66cf9609582421f44c4dc9833c08/client/db.go#L368)):*
done

---

*[roachpb/api.proto, line 559 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KDGR7JhTjQUogZffJL4:-KETnLWwN5QHsBztQVPa:2008002164) ([raw file](https://github.com/cockroachdb/cockroach/blob/47f82325faaa6329092d8c2a7d86ae3abf557049/roachpb/api.proto#L559)):*
I had to do that to eliminate the dependency from internal_raft.proto -> api.proto
This PR also adds a dependency from api.proto to internal_raft.proto

---

*[roachpb/internal_raft.proto, line 30 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KEOGUcLJZlynoqJGqQZ:-KETnfKMxW5j6pH3CVzR:2129497902) ([raw file](https://github.com/cockroachdb/cockroach/blob/7a85cc05b183e61e9224fcec7b8acfe4d09cfa1c/roachpb/internal_raft.proto#L30)):*
explained above

---

*[storage/client_raft_test.go, line 1564 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KEOGiI7voMgSfIBfNdP:-KEToWgqvaPbscSCoLUQ:-1820326818) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b3b8ee3212a66cf9609582421f44c4dc9833c08/storage/client_raft_test.go#L1564)):*
We don't have a systematic way to know when a command has executed on one of the replicas. I see Sleep() used in a few places in these tests. 

---

*[storage/replica_command.go, line 1657 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KEBNVkIm7ZtsSNeteiN:-KETovXjiiCR4nIxT05V:2118135454) ([raw file](https://github.com/cockroachdb/cockroach/blob/66db1b7c310585d1f7b583aed08032adcac3ea64/storage/replica_command.go#L1657)):*
The diff does tell you.

---

*[storage/replica_command.go, line 1557 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KEOHr5nrAWnOs6Auos1:-KETpdSWvvEd_QAcdnX3:1999346100) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b3b8ee3212a66cf9609582421f44c4dc9833c08/storage/replica_command.go#L1557)):*
Added a comment, but didn't rename

---

*[storage/replica_command.go, line 1654 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KEOCrsSNwxnKfcaQ1Ja:-KETwJcRoKphQHl7H989:-2145386761) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b3b8ee3212a66cf9609582421f44c4dc9833c08/storage/replica_command.go#L1654)):*
Done. Changed d.Key to roachpb.Key

---

*[storage/replica_test.go, line 4707 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KEOEgxVh3row5MQJoff:-KETvLZBDfYc10V431IA:537449651) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b3b8ee3212a66cf9609582421f44c4dc9833c08/storage/replica_test.go#L4707)):*
Done. I've added similar keys with varying values, and have also eliminated all the data generating code, and enumerate all the data for ease of readability.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Reviewed 11 of 11 files at r10.
Review status: all files reviewed at latest revision, 14 unresolved discussions.

---

*[storage/client_raft_test.go, line 1564 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KEOGiI7voMgSfIBfNdP:-KEUGiEYD8PlJdRI5GuY:-591616426) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b3b8ee3212a66cf9609582421f44c4dc9833c08/storage/client_raft_test.go#L1564)):*
Could you use a different key for the store 1 MVCCPut than you used for the replicated puts? Then the order of operations between them wouldn't matter and you wouldn't have to sleep.

If you do need to wait, it's better to use SucceedsWithin to read from the values instead of a fixed sleep. See `multiTestContext.waitForValues`.

---

*[storage/replica_command.go, line 1654 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KEOCrsSNwxnKfcaQ1Ja:-KEUGF1-JUxs0Oy7vINr:531056825) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b3b8ee3212a66cf9609582421f44c4dc9833c08/storage/replica_command.go#L1654)):*
Can't you remove the call to PrettyPrint too?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 13 of 15 files at r6, 9 of 11 files at r10.
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[client/batch.go, line 376 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KEOFoWZ5qjpP9svfbJX:-KEUuzh2suy5tEPpcL7g:547130398) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b3b8ee3212a66cf9609582421f44c4dc9833c08/client/batch.go#L376)):*
"It logs a diff of all the keys that are inconsistent when withDiff is set to true."

---

*[roachpb/api.proto, line 559 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KDGR7JhTjQUogZffJL4:-KEUuWRhgJ3hccN98PbE:-1242837539) ([raw file](https://github.com/cockroachdb/cockroach/blob/47f82325faaa6329092d8c2a7d86ae3abf557049/roachpb/api.proto#L559)):*
why do these dependencies matter? I thought removing c++ compilation of api.proto obviated those problems.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 11 files at r10.
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[storage/replica_command.go, line 1718 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KEW3BcsIHNRuVo8Zyz-:-KEW3BctDIJT9Gj2S6oU:1051221348) ([raw file](https://github.com/cockroachdb/cockroach/blob/f005eb5d39727d61c2cac4ea6fbf1fc4fd10096f/storage/replica_command.go#L1718)):*
this comment and the one above are misleading? what is it supposed to mean?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 2 of 13 files reviewed at latest revision, 11 unresolved discussions.

---

*[client/batch.go, line 376 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KEOFoWZ5qjpP9svfbJX:-KEWJ0zSF4l4Zhn38J27:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b3b8ee3212a66cf9609582421f44c4dc9833c08/client/batch.go#L376)):*
done

---

*[roachpb/api.proto, line 559 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KDGR7JhTjQUogZffJL4:-KEWJOAIjG61RAL4W8xd:-225903315) ([raw file](https://github.com/cockroachdb/cockroach/blob/47f82325faaa6329092d8c2a7d86ae3abf557049/roachpb/api.proto#L559)):*
VerifyChecksumRequest in api.proto now depends on RaftSnapshotData in internal_raft.proto. We need to add a api.proto -> internal_raft.proto dependency in this change, which means eliminating the internal_raft.proto -> api.proto dependency. Look at the changes at the top of both proto files.

---

*[storage/client_raft_test.go, line 1564 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KEOGiI7voMgSfIBfNdP:-KEWGSgzyXKygbR8eOVj:1546909141) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b3b8ee3212a66cf9609582421f44c4dc9833c08/storage/client_raft_test.go#L1564)):*
I've changed it to use a different key "e"

---

*[storage/replica_command.go, line 1654 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KEOCrsSNwxnKfcaQ1Ja:-KEWK55gtFL3dwQlxtl0:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b3b8ee3212a66cf9609582421f44c4dc9833c08/storage/replica_command.go#L1654)):*
done

---

*[storage/replica_command.go, line 1718 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KEW3BcsIHNRuVo8Zyz-:-KEWIVN7RohnHJctX-QE:-548754926) ([raw file](https://github.com/cockroachdb/cockroach/blob/f005eb5d39727d61c2cac4ea6fbf1fc4fd10096f/storage/replica_command.go#L1718)):*
Removed. I've added a few comments around. I hope it's easier to understand.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Reviewed 11 of 11 files at r12.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KE9neEblMFpKHvqmhO0:-KEX6gKP64_-YNlxIIeN:-1747511869) (no related file):*
@tamird any more comments on this. I'd like to land this and push to a Beta cluster so we know what's going on. Thanks

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 11 of 11 files at r11, 11 of 11 files at r12.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KE9neEblMFpKHvqmhO0:-KEXAKwBwVfEyYv4-FPM:677819321) (no related file):*
lgtm

---

*[client/batch.go, line 376 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KEOFoWZ5qjpP9svfbJX:-KEXA6-6qnYVVO3nqnYn:-1794885953) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b3b8ee3212a66cf9609582421f44c4dc9833c08/client/batch.go#L376)):*
now the tenses are different.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 2 of 13 files reviewed at latest revision, 9 unresolved discussions.

---

*[client/batch.go, line 376 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403#-KEOFoWZ5qjpP9svfbJX:-KEXGA46lko0QlQweOWB:-1180889527) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b3b8ee3212a66cf9609582421f44c4dc9833c08/client/batch.go#L376)):*
done as you want it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/5403)*
<!-- Sent from Reviewable.io -->
